diff --git a/include/psa/crypto_extra.h b/include/psa/crypto_extra.h
index 1f68925ff..2a5a1e7c8 100644
--- a/include/psa/crypto_extra.h
+++ b/include/psa/crypto_extra.h
@@ -499,11 +499,14 @@ psa_status_t mbedtls_psa_inject_entropy(const uint8_t *seed,
  * \retval #PSA_ERROR_NOT_SUPPORTED
  * \retval #PSA_ERROR_INSUFFICIENT_MEMORY
  */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wredundant-decls"
 psa_status_t psa_set_key_domain_parameters(psa_key_attributes_t *attributes,
                                            psa_key_type_t type,
                                            const uint8_t *data,
                                            size_t data_length);
 
+#pragma GCC diagnostic pop
 /**
  * \brief Get domain parameters for a key.
  *
diff --git a/library/psa_crypto_core.h b/library/psa_crypto_core.h
index 8c91b04d0..22662baf1 100644
--- a/library/psa_crypto_core.h
+++ b/library/psa_crypto_core.h
@@ -258,9 +258,12 @@ psa_status_t mbedtls_to_psa_error( int ret );
  * \return  The Mbed TLS cipher information of the cipher algorithm.
  *          \c NULL if the PSA cipher algorithm is not supported.
  */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wredundant-decls"
 const mbedtls_cipher_info_t *mbedtls_cipher_info_from_psa(
     psa_algorithm_t alg, psa_key_type_t key_type, size_t key_bits,
     mbedtls_cipher_id_t *cipher_id );
+#pragma GCC diagnostic pop
 
 /** Import a key in binary format.
  *
